homeassistant/components/rabbitair/fan.py

```diff
@@ -0,0 +1,148 @@
+            ]
+
+        self._attr_speed_count = len(SPEED_LIST)
+
+        self._get_state_from_coordinator_data()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._get_state_from_coordinator_data()
+        super()._handle_coordinator_update()
+
+    def _get_state_from_coordinator_data(self) -> None:
+        """Populate the entity fields with values from the coordinator data."""
+        data = self.coordinator.data
+
+        # Power on/off
+        self._attr_is_on = data.power
+
+        # Speed as a percentage
+        if data.speed is None:
+            self._attr_percentage = None
+        elif data.speed is Speed.SuperSilent:
+            self._attr_percentage = 0
+        else:
+            self._attr_percentage = ordered_list_item_to_percentage(
+                SPEED_LIST, data.speed
+            )
+
+        # Preset mode
+        if data.mode is None:
+            self._attr_preset_mode = None
+        else:
+            # Get key by value in dictionary
+            self._attr_preset_mode = next(
+                k for k, v in PRESET_MODES.items() if v == data.mode
+            )
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set new preset mode."""
+        await self._set_state(mode=PRESET_MODES[preset_mode])
+        self._attr_preset_mode = preset_mode
+        self.async_write_ha_state()
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed of the fan, as a percentage."""
+        value = percentage_to_ordered_list_item(SPEED_LIST, percentage)
+        await self._set_state(speed=value)
+        self._attr_percentage = percentage
+        self.async_write_ha_state()
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        mode_value: Mode | None = None
+        if preset_mode is not None:
+            mode_value = PRESET_MODES[preset_mode]
+        speed_value: Speed | None = None
+        if percentage is not None:
+            speed_value = percentage_to_ordered_list_item(SPEED_LIST, percentage)
+        await self._set_state(power=True, mode=mode_value, speed=speed_value)
+        self._attr_is_on = True
+        if percentage is not None:
+            self._attr_percentage = percentage
+        if preset_mode is not None:
+            self._attr_preset_mode = preset_mode
+        self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the fan off."""
+        await self._set_state(power=False)
+        self._attr_is_on = False
+        self.async_write_ha_state()
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if device is on."""
+        return self._attr_is_on
```

Comment:
I think @frenck didn't realize `FanEntity` ignores `_attr_is_on`. It's somewhat confusing to use `_attr_is_on` here and then also overriding the property of the base class.

Furthermore, the base class implements `is_on` like this:
```py
    @property
    def is_on(self) -> bool | None:
        """Return true if the entity is on."""
        return (
            self.percentage is not None and self.percentage > 0
        ) or self.preset_mode is not None
```

The custom implementation in this class will allow the fan to report a fan speed which is not None or 0 or a preset mode which is not None, yet the state will be `off`.
This seems illogical, why doesn't the `is_on` from the base class suffice?

@bdraco could you comment?

Reply:
is_on should probably be marked as @final in the base fan entity

Reply:
OK, so if the API has a separate flag for if the fan is on, that should be used to conditionally set the percentage and preset_mode to 0 / None.

Reply:
> is_on should probably be marked as https://github.com/Final in the base fan entity

That's easier said than done, it seems about 2/3 of the fan entities override `is_on`.

Reply:
Changed the logic to use `is_on` from the base class. Now power off will cause the percentage and preset_mode to be set to None. Setting the percentage to 0 will cause the power to be turned off. Setting the percentage > 0 or preset_mode to some value will cause the unit to turn on.